### ISSUE TITLE
RHEL/CentOS 8 missing net46/Microsoft.NET.Sdk.Razor.Tasks.dll

### ISSUE THREAD

**Issue by michaelbarkdoll** [dotnet/source-build#2006]
Hi, I have an issue with a missing .dll file.

My C# extension no longer loads in VS Code due to this missing dll.

OmniSharp/omnisharp-vscode#4360

On the Ubuntu 20.04 LTS the dotnet-sdk-5.0 package has the following dll files:
/usr/share/dotnet/sdk/5.0.102/Sdks/Microsoft.NET.Sdk.Razor/tasks/net46/Microsoft.NET.Sdk.Razor.Tasks.dll
/usr/share/dotnet/sdk/5.0.102/Sdks/Microsoft.NET.Sdk.Razor/tasks/net5.0/Microsoft.NET.Sdk.Razor.Tasks.dll

However, my CentOS 8 only has the following file installed after upgrading to 5.0.102:
/usr/lib64/dotnet/sdk/5.0.102/Sdks/Microsoft.NET.Sdk.Razor/tasks/net5.0/Microsoft.NET.Sdk.Razor.Tasks.dll

So it appears that in RHEL/CentOS 8 the following file(s) are missing:
/usr/lib64/dotnet/sdk/5.0.102/Sdks/Microsoft.NET.Sdk.Razor/tasks/net46/Microsoft.NET.Sdk.Razor.Tasks.dll

ls -la tasks/net46/
-rw-r--r-- 1 root root 77192 Dec 14 20:17 Microsoft.NET.Sdk.Razor.Tasks.dll
-rw-r--r-- 1 root root 189832 Dec 14 20:17 System.Collections.Immutable.dll
-rw-r--r-- 1 root root 464776 Dec 14 20:17 System.Reflection.Metadata.dll

**Comment by dagood** [comment]
For context: CentOS builds .NET from source, based on the infrastructure in this repo dotnet/source-build. That's why it's different from the build produced by Microsoft.

We currently have a patch that removes the `net46` build for this part of the SDK, so we'll have to look into removing the patch to build the `net46` files. https://github.com/dotnet/source-build/blob/release/5.0/patches/aspnetcore/0010-Don-t-build-for-full-fx.patch. (It's not new to 5.0.102, it was in 5.0.100 as well: https://github.com/dotnet/source-build/blob/v5.0.100-SDK/patches/aspnetcore/0012-Don-t-build-for-full-fx.patch.)
@crummel do you happen to know why this patch was needed?

/cc @omajid (FWIW, the test script at https://github.com/dotnet/source-build/issues/1950 seems to work fine with CentOS 8's 5.0.102.)

**Comment by omajid** [comment]
> @omajid (FWIW, the test script at #1950 seems to work fine with CentOS 8's 5.0.102.)

I took a quick look and it looks like that script only tries out a `dotnet new console` project, which works with omnisharp. Many other project types (`web`, `blazorserver`) don't.

**Comment by dagood** [comment]
We got an error message from #2008 showing what the patch was avoiding:

> Based on this error:
> 
> > /src/artifacts/src/aspnetcore.2670c128d522473e146ff9f8159bfffdfe694cd9/src/Razor/Microsoft.NET.Sdk.Razor/src/Microsoft.NET.Sdk.Razor.csproj : error NU1202: Package Microsoft.Build.Framework 16.8.3 is not compatible with net46 (.NETFramework,Version=v4.6). Package Microsoft.Build.Framework 16.8.3 supports: netstandard2.0 (.NETStandard,Version=v2.0) [/src/.dotnet/sdk/5.0.100/NuGet.targets]
> 
> I would guess we need to get rid of this `DotNetBuildFromSource` condition:
> 
> https://github.com/dotnet/msbuild/blob/39993bd9d02917993b6147a9973dc3aa6e9b00c7/src/Directory.Build.props#L29-L30
> 
> ```
>     <LibraryTargetFrameworks>$(FullFrameworkTFM);netstandard2.0</LibraryTargetFrameworks>
>     <LibraryTargetFrameworks Condition="'$(DotNetBuildFromSource)' == 'true'">netstandard2.0</LibraryTargetFrameworks>
> ```

**Comment by tvildo** [comment]
On Centos 33 Omnisharp was crashing with exception and to solve that I had to install additional package:
dotnet add package Microsoft.NET.Sdk.Razor --version 3.1.12

**Comment by omajid** [comment]
Hey folks. Any progress on this? Is there anything I can do to help here?

**Comment by richlander** [comment]
FYI: https://bugs.archlinux.org/task/70018

**Comment by crummel** [comment]
We have a planning/triage meeting on Monday, I'll make sure to bring this up as something to prioritize.  I'll noodle with the MSBuild issue a bit, I know we had to change that patch for the feature branch upgrade so maybe we can fix that now.

**Comment by tmds** [comment]
I just hit this one too.

```
[fail]: OmniSharp.MSBuild.ProjectLoader
        The "Microsoft.AspNetCore.Razor.Tasks.DiscoverDefaultScopedCssItems" task could not be loaded from the assembly /usr/lib64/dotnet/sdk/5.0.202/Sdks/Microsoft.NET.Sdk.Razor/build/netstandard2.0/../../tasks/net46/Microsoft.NET.Sdk.Razor.Tasks.dll. Invalid Image Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask.
[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file '/tmp/web/web.csproj'.
```

**Comment by zpxp** [comment]
Is there a workaround for this issue? It makes the IDE (via omnisharp) unusable.

**Comment by exyi** [comment]
@zpxp 
> Is there a workaround for this issue? It makes the IDE (via omnisharp) unusable.

Adding the reference to Microsoft.NET.Sdk.Razor as was suggested: `dotnet add package Microsoft.NET.Sdk.Razor --version 3.1.12`. At least it fixed the issue for me on Arch Linux.

**Comment by victormikael** [comment]
> @zpxp
> 
> > Is there a workaround for this issue? It makes the IDE (via omnisharp) unusable.
> 
> Adding the reference to Microsoft.NET.Sdk.Razor as was suggested: `dotnet add package Microsoft.NET.Sdk.Razor --version 3.1.12`. At least it fixed the issue for me on Arch Linux.

Thanks! Work for me.

I too try install bin packages `dotnet-sdk-bin aspnet-runtime-bin aspnet-targeting-pack-bin` from AUR and worked. To use it not is needed to add Razor package old.

**Comment by omajid** [comment]
Hey @dleeapho can we increase the priority of this issue and it's short-term and long-term fixes too? Due to this bug, the IDE experience (auto-completion/Intellisense) is completely broken for ASP.NET projects in 5.0.

**Comment by crummel** [comment]
I'm looking at this now, I'm hoping to get it done this week.

**Comment by zpxp** [comment]
As another workaround, you can use the vscode dev containers and docker to develop inside the .NET 5 docker template image

**Comment by JoeRobich** [comment]
From the VS Code C# extension side of things, OmniSharp (the C# language server) runs on Full Framework (Mono or .NET Framework) to support the greatest number of project types. This means when the Web SDK targets are loading MSBuild tasks, it is trying to load the Full Framework build of the assemblies (net46 or net472) instead of the .NET 5 assemblies shipped with the source build version of teh .NET SDK.

**Comment by 2bdkid** [comment]
Also running into this issue

**Comment by omajid** [comment]
This was fixed by https://github.com/dotnet/source-build/pull/2178 (thanks, @crummel!) and included in the latest release: https://github.com/dotnet/source-build/releases/tag/v5.0.205-SDK

AFAIK, updates have gone out for CentOS 8, Fedora, and RHEL.

I see that the Arch Linux package has been updated too: https://github.com/archlinux/svntogit-community/commit/055b029edd458418c5f756f62f9691202b47d354#diff-3e341d2d9c67be01819b25b25d5e53ea3cdf3a38d28846cda85a195eb9b7203a

Is there anything else to do here? Can we close this issue now?

**Comment by ultimaweapon** [comment]
I can confirm the issue has been fixed on Arch Linux with the latest version.

**Comment by crummel** [comment]
Closing this issue then, thanks for testing!

Fixed by #2178.
